README.md
# gulp-amd-wrap
![Language](https://img.shields.io/badge/-TypeScript-blue.svg)
[![Build Status](https://travis-ci.org/searchfe/gulp-amd-wrap.svg?branch=master)](https://travis-ci.org/searchfe/gulp-amd-wrap)
[![Coveralls](https://img.shields.io/coveralls/searchfe/gulp-amd-wrap.svg)](https://coveralls.io/github/searchfe/gulp-amd-wrap)
[![npm package](https://img.shields.io/npm/v/gulp-amd-wrap.svg)](https://www.npmjs.org/package/gulp-amd-wrap)
[![npm downloads](http://img.shields.io/npm/dm/gulp-amd-wrap.svg)](https://www.npmjs.org/package/gulp-amd-wrap)
[![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)

gulp-amd-wrap是一个分析amd模块，并进行预编译处理的的gulp插件，主要完成根据依赖分析及项目路径生成模块声明及引用的moduleID，并封装成amd规范的模块。

## Install

```bash
npm i gulp-amd-wrap --save-dev
```

## Example

```Typescript
import { amdWrap } from 'gulp-amd-wrap';

gulp.src(
  // 资源
  `${__dirname}\/assert/*.js`, {
      // 工程baseUrl
      base: __dirname,
    },
).pipe(amdWrap({
  baseUrl: '/assert/',
  prefix: 'wiseindex/',
  // 不参与amd-hook分析的文件
  exelude: ['/exclude-**.js', '/dist/**'],
  alias: [{
    moduleId: 'debug',
    path: 'util/debug',
    prifix: false
  }]
})).pipe(
  gulp.dest(`${__dirname}\/dist\/`),
);

```

```javascript
// Before
define(function() {
  'use strict';

  var fishObj = {
    name: 'nimo'
  };
  var moduleA = require('@scope/moduleA');
  var moduleB = require('./moduleB');

  require(['A', './B', '/C', '@D/E', './util/debug'], function(a, b, c, d){
    console.log(a, b, c, d);
  });

  return fishObj;

});

// After
define('wiseindex/assert/minify-define', [
    'require',
    '@scope/moduleA',
    'wiseindex/assert/moduleB'
], function (require, moduleA, moduleB) {
    'use strict';
    var fishObj = { name: 'nimo' };
    require([
        'A',
        'wiseindex/assert/B',
        '/C',
        '@D/E',
        'debug'
    ], function (a, b, c, d) {
        console.log(a, b, c, d);
    });
    return fishObj;
});

```


## API

[API DOC](https://searchfe.github.io/gulp-amd-wrap/)


CHANGELOG.md
## [1.1.2](https://github.com/searchfe/gulp-amd-wrap/compare/v1.1.1...v1.1.2) (2019-11-08)


### Bug Fixes

* useMd5设置后moduleID未更改 ([67621ce](https://github.com/searchfe/gulp-amd-wrap/commit/67621ce))

## [1.1.1](https://github.com/searchfe/gulp-amd-wrap/compare/v1.1.0...v1.1.1) (2019-11-08)


### Bug Fixes

* useMd5设置后moduleID未更改 ([8c0efbb](https://github.com/searchfe/gulp-amd-wrap/commit/8c0efbb))

# [1.1.0](https://github.com/searchfe/gulp-amd-wrap/compare/v1.0.10...v1.1.0) (2019-10-23)


### Bug Fixes

* alias bug ([fd2eeaf](https://github.com/searchfe/gulp-amd-wrap/commit/fd2eeaf))
* never push function argumentds into deps array ([5787bcc](https://github.com/searchfe/gulp-amd-wrap/commit/5787bcc))
* never push function argumentds into deps array ([49c3750](https://github.com/searchfe/gulp-amd-wrap/commit/49c3750))
* test for keep var ([fb3c073](https://github.com/searchfe/gulp-amd-wrap/commit/fb3c073))


### Features

* add alias ([f41e0e1](https://github.com/searchfe/gulp-amd-wrap/commit/f41e0e1))
* eslint upgrade ([483d3e4](https://github.com/searchfe/gulp-amd-wrap/commit/483d3e4))
* support cache ([a41ee37](https://github.com/searchfe/gulp-amd-wrap/commit/a41ee37))
* support cache ([730a2d5](https://github.com/searchfe/gulp-amd-wrap/commit/730a2d5))
* upgrade predeps ([ef62705](https://github.com/searchfe/gulp-amd-wrap/commit/ef62705))
* upgrade preinstall ([2e08a28](https://github.com/searchfe/gulp-amd-wrap/commit/2e08a28))
* upgrade preinstall ([fd9fa34](https://github.com/searchfe/gulp-amd-wrap/commit/fd9fa34))
* upgrade preinstall ([894d3bb](https://github.com/searchfe/gulp-amd-wrap/commit/894d3bb))

## [1.0.10](https://github.com/searchfe/gulp-amd-wrap/compare/v1.0.9...v1.0.10) (2019-09-19)


### Bug Fixes

* support require dynamic module ([e5836ac](https://github.com/searchfe/gulp-amd-wrap/commit/e5836ac))
* support require dynamic module ([ac4eb0c](https://github.com/searchfe/gulp-amd-wrap/commit/ac4eb0c))

## [1.0.9](https://github.com/searchfe/gulp-amd-wrap/compare/v1.0.8...v1.0.9) (2019-09-19)


### Bug Fixes

* support require dynamic module ([0bc28e4](https://github.com/searchfe/gulp-amd-wrap/commit/0bc28e4))
* support require dynamic module ([38d0349](https://github.com/searchfe/gulp-amd-wrap/commit/38d0349))

## [1.0.8](https://github.com/searchfe/gulp-amd-wrap/compare/v1.0.7...v1.0.8) (2019-09-19)


### Bug Fixes

* support require dynamic module ([1e37860](https://github.com/searchfe/gulp-amd-wrap/commit/1e37860))

## [1.0.7](https://github.com/searchfe/gulp-amd-wrap/compare/v1.0.6...v1.0.7) (2019-07-29)

### Bug Fixes

* no elements in parser ([46823b3](https://github.com/searchfe/gulp-amd-wrap/commit/46823b3))

## [1.0.6](https://github.com/searchfe/gulp-amd-wrap/compare/v1.0.5...v1.0.6) (2019-07-04)


### Bug Fixes

* add callexpression require ([2420ce9](https://github.com/searchfe/gulp-amd-wrap/commit/2420ce9))
* add callexpression require ([fd8b55a](https://github.com/searchfe/gulp-amd-wrap/commit/fd8b55a))

## [1.0.5](https://github.com/searchfe/gulp-amd-wrap/compare/v1.0.4...v1.0.5) (2019-06-05)


### Bug Fixes

* trigger patch ([444395e](https://github.com/searchfe/gulp-amd-wrap/commit/444395e))

## [1.0.4](https://github.com/searchfe/gulp-amd-wrap/compare/v1.0.3...v1.0.4) (2019-06-05)

### Bug Fixes

* fix semantic config bug ([fb75b9e](https://github.com/searchfe/gulp-amd-wrap/commit/fb75b9e))


src/parser.ts
/*
 * @Author: qiansc
 * @Date: 2019-04-28 14:43:21
 * @Last Modified by: qiansc
 * @Last Modified time: 2019-10-23 13:13:05
 */
import { generate } from 'escodegen';
import { parse, parseScript } from 'esprima';
import { traverse } from 'estraverse';
import { existsSync } from 'fs';
import { AsyncAnalyzer } from './async-analyzer';
import { DependencyAnalyzer } from './dependency-analyzer';
import { include } from './filter';
import { parseAbsolute, parseBase, aliasConf } from './moduleID';
import { basename, dirname, extname, resolve } from 'path';
const md5File = require('md5-file');

export class Parser {
  private cwd: string;
  private ast;
  private myModuleId: string;
  private parseDefine;
  private dependences: string[];
  constructor (
    private contents: Buffer,
    private filePath: string,
    private root: string,
    private prefix: string,
    private alias?: aliasConf[],
    private staticBaseUrl?: string) {
      this.cwd = dirname(filePath);
      this.parse();
      this.dependences = [];
      this.parseDefine = 0;
  }
  public getContent () {
      return this.contents;
  }
  public getModuleId () {
      return this.myModuleId;
  }
  public getDependences () {
      return Array.from(new Set(this.dependences));
  }
  public isHasObj (arr, val) {
      let flag = false;
      if (!arr || arr.length === 0) { return false; }
      arr.forEach((element) => {
          if (JSON.stringify(element).indexOf(JSON.stringify(val)) !== -1) {
              flag = true;
          }
      });
      return flag;
  }
  public hook (hookOption: HookOption = {}) {
      /** 生成的ModuleId md5后缀来避免其他模块引用 @molecule/toptip2_134dfas */
      this.parseDefine = 0;
      let md5Value: string = '';

      if (hookOption.useMd5 === true || (hookOption.useMd5 && hookOption.useMd5.useMd5)) {
          const exlude = hookOption.useMd5.exlude; // 要被排除的文件

          if (!include(resolve(this.filePath), exlude, this.root)) {
              // 不在md5排除名单中
              try {
                  md5Value = '_' + md5File.sync(this.filePath.replace('.js', '.ts')).slice(0, 7);
              } catch (e) {
                  console.log(e);
              }
          }
      }
      // console.log(this.contents.toString());
      traverse(this.ast, {
          enter: (node) => {
              if (node.type === 'CallExpression') {
                  // 如果是define
                  if (node.callee && node.callee.name === 'define') {
                      this.parseDefine++;
                      if (this.parseDefine === 1) {
                          // 首参数是function，推入依赖数组
                          if (node.arguments[0].type === 'FunctionExpression') {
                              node.arguments.unshift(
                                  { type: 'ArrayExpression', elements: [] }
                              );
                          }
                          // 首参数推入moduleId => define("@molecule/toptip/main",xxxx)
                          if (node.arguments[0].type === 'ArrayExpression') {
                              node.arguments.unshift(
                                  {
                                      type: 'Literal',
                                      value: parseBase(this.root, this.filePath, this.prefix, this.alias) + md5Value
                                  }
                              );
                          } else if (node.arguments[0].type === 'ObjectExpression' || node.arguments[0].type === 'Identifier') {
                              node.arguments.unshift({ type: 'ArrayExpression', elements: [] });
                              node.arguments.unshift({
                                  type: 'Literal',
                                  value: parseBase(this.root, this.filePath, this.prefix, this.alias) });
                          } else if (node.arguments[0].type === 'Literal') {
                              const moduleId = node.arguments[0].value;
                              if (moduleId.split('/').pop() === basename(this.filePath, extname(this.filePath))) {
                                  const prefix = moduleId.match(/^\./) === null ? '' : this.prefix;
                                  node.arguments[0].value = parseBase(this.root, this.filePath, prefix, this.alias);
                              }
                          }
                          this.myModuleId = node.arguments[0].value;
                          const analyzer = new DependencyAnalyzer(
                              this.cwd,
                              node,
                              this.prefix,
                              this.alias,
                              this.root,
                              this.staticBaseUrl);

                          const deps = analyzer.analysis(
                              (dep, requireNode, parent) => {
                                  // requireNode.id.name = {type};
                                  // console.log('analyzer', dep, requireNode);
                              }
                          );

                          // 第二参数是依赖数组 => define("", ['require', 'exports', 'md5-file'])
                          if (node.arguments[1].elements) {
                              node.arguments[1].elements.forEach((element, index) => {
                                  const valueString = element.value;
                                  /** depPath: 实际依赖的相对路径文件。如果是node_module就为空 */
                                  const depPath = parseAbsolute(dirname(this.filePath), valueString + '.ts');
                                  if (existsSync(depPath)) {
                                      let md5 = '';
                                      if (hookOption.useMd5 === true ||
                                        (hookOption.useMd5 && hookOption.useMd5.useMd5)) {
                                          const exlude = hookOption.useMd5.exlude; // 要被排除的文件

                                          if (!include(resolve(this.filePath), exlude, this.root)) {
                                              // 不在md5排除名单中
                                              try {
                                                  md5 = '_' + md5File.sync(depPath).slice(0, 7);
                                              } catch (e) {
                                                  console.log(e);
                                              }
                                          }
                                      }
                                      // moduleid 示例：@molecule/toptip/main_dc85e717d6352fa285bc70bc2d1d3595
                                      const moduleid = parseBase(this.root, depPath, this.prefix, this.alias) + md5;
                                      console.log(moduleid);
                                      node.arguments[1].elements[index].value = moduleid;
                                  }
                              });
                          }

                          if (node.arguments[1] && node.arguments[1].elements &&
                !this.isHasObj(node.arguments[1].elements, 'require')) {
                              node.arguments[1].elements.unshift({ type: 'Literal', value: 'require' });
                          }
                          if (node.arguments[2] && node.arguments[2].params) {
                              if (!this.isHasObj(node.arguments[2].params, 'require')) {
                                  node.arguments[2].params.unshift({ type: 'Identifier', name: 'require' });
                              }
                              node.arguments[2].params.forEach((item, index) => {
                                  if (item.name === 'exports') {
                                      node.arguments[1].elements[index] = { type: 'Literal', value: 'exports' };
                                  } else if (item.name === 'module') {
                                      node.arguments[1].elements[index] = { type: 'Literal', value: 'module' };
                                  }
                              });
                          }
                          deps.forEach((dep) => {
                              if (node.arguments[1] && node.arguments[1].elements &&
                  node.arguments[1].elements.map((e) => e.value).indexOf(dep.moduleID) < 0) {
                                  node.arguments[1].elements.push({ type: 'Literal', value: dep.moduleID });
                                  // factory函数的参数中推入依赖对应的变量名
                                  // if (dep.name) {
                                  //   node.arguments[2].params.push({ type: 'Identifier', name: dep.name });
                                  // }
                                  this.dependences.push(dep.moduleID);
                              }
                          });
                          if (hookOption.removeModuleId) {
                              node.arguments.shift();
                          }
                      }
                  }
                  if (node.callee && node.callee.name === 'require' && this.parseDefine < 2) {
                      const firstArg = node.arguments[0];
                      // 首参数是依赖数组，推入moduleId
                      if (firstArg.type === 'ArrayExpression') {
                          firstArg.elements.map((element) => {
                              if (element.value && element.value.match(/^\.\.?\//) !== null) {
                                  element.value = parseBase(
                                      this.root,
                                      parseAbsolute(this.cwd, element.value),
                                      this.prefix,
                                      this.alias);
                                  element.raw = `"${element.value}"`;
                                  this.dependences.push(element.value);
                              }
                          });
                      }
                      if (firstArg.type === 'Literal') {
                          if (firstArg.value && firstArg.value.match(/^\.\.?\//) !== null) {
                              const baseUrl = extname(firstArg.value) !== '.json' ? this.root : this.staticBaseUrl || this.root;
                              firstArg.value = parseBase(
                                  baseUrl,
                                  parseAbsolute(this.cwd, firstArg.value),
                                  this.prefix,
                                  this.alias);
                              firstArg.raw = `"${firstArg.value}"`;
                              this.dependences.push(firstArg.value);
                          }
                      }
                  }
              }
              return node;
          },
          leave: (node) => {
              if (node.type === 'CallExpression') {
                  // 如果是define
                  if (node.callee && node.callee.name === 'define') {
                      this.parseDefine--;
                  }
              }
              // const aa = new AsyncAnalyzer(
              //   this.cwd,
              //   node,
              //   this.root,
              // );
              // aa.analysis();
          }
      });
      this.contents = generate(this.ast);
      return this;
  }
  private parse () {
      this.ast = parseScript(this.contents.toString());
  }

  /**
   * define(function() {});
   */
  private isSingelFunction () {
      // do nothing
  }

    /**
   * define([...], function() {});
   */

    /**
   * define('XXX', [...], function() {});
   */
}

interface HookOption {
  removeModuleId?: boolean;
  useMd5?: any;
}
